examples/fakerealtime/main.go

```diff
@@ -0,0 +1,225 @@
+        primary key (update_id, channel)
+    )`,
+
+		`drop table if exists fakerealtime.messages`,
+		`create table fakerealtime.messages (
+        channel string,
+        msg_id int,
+        message string,
+        primary key (channel, msg_id)
+    )`,
+	}
+	for _, stmt := range statements {
+		if _, err := db.Exec(stmt); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type statistics struct {
+	sync.Mutex
+	writeTimes stats.Float64Data
+}
+
+func (s *statistics) recordWrite(start time.Time) {
+	duration := time.Now().Sub(start)
+	s.Lock()
+	defer s.Unlock()
+	s.writeTimes = append(s.writeTimes, float64(duration.Nanoseconds()))
+}
+
+func (s *statistics) report() {
+	for range time.Tick(time.Second) {
+		s.Lock()
+		writeTimes := s.writeTimes
+		s.writeTimes = nil
+		s.Unlock()
+
+		// The stats functions return an error only when the input is empty.
+		mean, _ := stats.Mean(writeTimes)
+		stddev, _ := stats.StandardDeviation(writeTimes)
+		log.Infof("wrote %d messages, latency mean=%s, stddev=%s",
+			len(writeTimes), time.Duration(mean), time.Duration(stddev))
+	}
+}
+
+type writer struct {
+	db          *sql.DB
+	numChannels int
+	wg          *sync.WaitGroup
+	stats       *statistics
+}
+
+func (w writer) run() {
+	defer w.wg.Done()
+	for {
+		if err := w.writeMessage(); err != nil {
+			log.Errorf("error writing message: %s", err)
+		}
+	}
+}
+
+func (w writer) writeMessage() error {
+	start := time.Now()
+	defer w.stats.recordWrite(start)
+	channel := fmt.Sprintf("room-%d", rand.Int31n(int32(w.numChannels)))
+	message := start.String()
+
+	// TODO(bdarnell): retry only on certain errors.
+	for {
+		txn, err := w.db.Begin()
+		if err != nil {
+			continue
+		}
+
+		// TODO(bdarnell): make this a subquery when subqueries are supported on insert.
+		row := txn.QueryRow(`select max(msg_id) from fakerealtime.messages where channel=$1`, channel)
+		var maxMsgID sql.NullInt64
+		if err := row.Scan(&maxMsgID); err != nil {
+			_ = txn.Rollback()
+			continue
+		}
+		if !maxMsgID.Valid {
+			maxMsgID.Int64 = 0
+		}
+		newMsgID := maxMsgID.Int64 + 1
+
+		row = txn.QueryRow(`select max(update_id) from fakerealtime.updates`, channel)
+		var maxUpdateID sql.NullInt64
+		if err := row.Scan(&maxUpdateID); err != nil {
+			_ = txn.Rollback()
+			continue
+		}
+		if !maxUpdateID.Valid {
+			maxUpdateID.Int64 = 0
+		}
+		newUpdateID := maxUpdateID.Int64 + 1
+
+		if _, err := txn.Exec(`insert into fakerealtime.messages (channel, msg_id, message) values ($1, $2, $3)`,
+			channel, newMsgID, message); err != nil {
+			_ = txn.Rollback()
+			continue
+		}
+
+		if _, err := txn.Exec(`insert into fakerealtime.updates (update_id, channel, msg_id) values ($1, $2, $3)`,
+			newUpdateID, channel, newMsgID); err != nil {
+			_ = txn.Rollback()
+			continue
+		}
+
+		if err := txn.Commit(); err == nil {
+			return nil
+		}
+	}
+}
+
+func main() {
+	var dbDriver string
+	flag.StringVar(&dbDriver, "db-driver", "cockroach", "database driver to use")
```

Comment:
Yeah, I'm attempting to be more flexible, although I haven't tested any of this. If we want our examples to be flexible we should probably have some shared setup code; if we're going to duplicate everything in each self-contained example we should probably just keep it minimal.